File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int maxLength = 0;
        int maxFreq = 0;
        int[] charArr = new int[26];

        for(int right = 0; right < s.length(); right++) {
            charArr[s.charAt(right) - 'A']++;

            maxFreq = Math.max(maxFreq, charArr[s.charAt(right) - 'A']);

            while((right - left + 1) - maxFreq > k) {
                charArr[s.charAt(left) - 'A']--;
                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
        
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.rs
use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut last_seen: HashMap<char, usize> = HashMap::new();
        let mut left = 0;
        let mut max_len = 0;

        for (right, ch) in s.chars().enumerate() {
            if let Some(&prev_index) = last_seen.get(&ch) {
                left = left.max(prev_index + 1);
            }

            last_seen.insert(ch, right);

            let current_len = right - left + 1;
            max_len = max_len.max(current_len);
        }

        max_len as i32
    }
}